File: app/src/main/java/com/nutrition/express/common/CommonRVAdapter.java
package com.nutrition.express.common;

import android.graphics.Color;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.nutrition.express.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;


/**
 * Created by huang on 10/19/16.
 */

public class CommonRVAdapter extends RecyclerView.Adapter<CommonViewHolder> {
    private static final int TYPE_DATA_BASE = 1000;
    private static final int TYPE_UNKNOWN = 0;       //未知类型
    /* 状态 */
    private static final int EMPTY = 10;                //显示EMPTY VIEW
    private static final int LOADING = 11;              //显示LOADING VIEW
    private static final int LOADING_FAILURE = 12;      //显示LOADING FAILURE VIEW
    private static final int LOADING_NEXT = 13;         //显示LOADING NEXT VIEW
    private static final int LOADING_NEXT_FAILURE = 14; //显示LOADING NEXT FAILURE VIEW
    private static final int LOADING_FINISH = 15;       //显示LOADING FINISH VIEW

    private int state = LOADING_FINISH;

    //保存了layout_id与MultiType键值对
    private SparseArray<MultiType> typeArray;
    //保存了数据类型名称与layout_id的键值对
    private HashMap<String, Integer> typeMap;
    private OnLoadListener loadListener;
    private List<Object> data = new ArrayList<>();

    private View.OnClickListener onRetryListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            if (loadListener != null) {
                loadListener.retry();
                if (data.size() > 0) {
                    state = LOADING_NEXT;
                } else {
                    state = LOADING;
                }
                notifyItemChanged(data.size());
            }
        }
    };

    private CommonRVAdapter(Builder builder) {
        typeArray = builder.typeArray;
        typeMap = builder.typeMap;
        loadListener = builder.loadListener;
        if (builder.data == null) {
            state = LOADING;
        } else if (builder.data.length == 0) {
            state = EMPTY;
        } else {
            Collections.addAll(data, builder.data);
        }
    }

    @Override
    public CommonViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        MultiType type = typeArray.get(viewType);
        if (type == null) {     //check if unknown type
            TextView textView = (TextView) inflater
                    .inflate(android.R.layout.simple_list_item_1, parent, false);
            textView.setText("Error!!! Unknown type!!!");
            textView.setTextColor(Color.parseColor("#ff0000"));
            return new CommonViewHolder(textView);
        } else {
            View view = inflater.inflate(type.layout, parent, false);
            if (state == LOADING_FAILURE) {
                view.setOnClickListener(onRetryListener);
            } else if (state == LOADING_NEXT_FAILURE) {
                view.setOnClickListener(onRetryListener);
            }
            return type.creator.createVH(view);
        }
    }

    @Override
    public void onBindViewHolder(CommonViewHolder holder, int position) {
        if (position < data.size()) {
            holder.bindView(data.get(position));
        } else {
            holder.bindView(null);
            if (state == LOADING_NEXT) {
                loadListener.loadNextPage();
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == data.size()) {
            return state;
        } else {
            Integer type = typeMap.get(data.get(position).getClass().getName());
            return (null == type ? TYPE_UNKNOWN : type);
        }
    }

    @Override
    public int getItemCount() {
        if (data.size() > 0) {
            if (state == EMPTY) {
                state = LOADING_FINISH;
            }
        } else if (state == LOADING_FINISH) {
            state = EMPTY;
        }
        return state == LOADING_FINISH ? data.size() : data.size() + 1;
    }

    /**
     *
     * @param data 列表数据，为空或无数据则表示加载完毕，不再自动加载下一页无论autoLoadingNext是否为true，
     * @param autoLoadingNext 是否自动加载下一页
     */
    public void append(Object[] data, boolean autoLoadingNext) {
        if (data != null && data.length > 0) {
            int lastDataIndex = this.data.size();
            int size = data.length;
            Collections.addAll(this.data, data);
            if (autoLoadingNext) {
                state = LOADING_NEXT;
                size++;
            } else {
                state = LOADING_FINISH;
            }
            notifyItemRangeChanged(lastDataIndex, size);
        } else {
            if (this.data.size() > 0) {
                //all data has been loaded
                state = LOADING_FINISH;
            } else {
                //no data, show empty view
                state = EMPTY;
            }
            notifyItemChanged(this.data.size());
        }
    }

    public void showLoadingFinish() {
        state = LOADING_FINISH;
        notifyItemChanged(data.size());
    }

    /**
     * 目前无数据，显示加载失败
     */
    public void showLoadingFailure() {
        state = LOADING_FAILURE;
        notifyItemChanged(data.size());
    }

    /**
     * 目前有数据，显示加载下一页失败
     */
    public void showLoadingNextFailure() {
        state = LOADING_NEXT_FAILURE;
        notifyItemChanged(data.size());
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public interface OnLoadListener {
        void retry();
        void loadNextPage();
    }

    public interface CreateViewHolder {
        CommonViewHolder createVH(View view);
    }

    private static class MultiType {
        int layout;
        CreateViewHolder creator;

        private MultiType(int layout, CreateViewHolder creator) {
            this.layout = layout;
            this.creator = creator;
        }
    }

    public static class Builder {
        private int emptyView = 0;
        private int loadingView = 0;
        private int failureView = 0;
        private int nextView = 0;
        private int nextFailureView = 0;
        private CreateViewHolder emptyCreator = null;
        private CreateViewHolder loadingCreator = null;
        private CreateViewHolder failureCreator = null;
        private CreateViewHolder nextCreator = null;
        private CreateViewHolder nextFailureCreator = null;
        private OnLoadListener loadListener;
        private Object[] data;
        private SparseArray<MultiType> typeArray = new SparseArray<>();
        private HashMap<String, Integer> typeMap = new HashMap<>();
        private int base = TYPE_DATA_BASE;
        private CreateViewHolder defaultCreator = new CreateViewHolder() {
            @Override
            public CommonViewHolder createVH(View view) {
                return new CommonViewHolder(view);
            }
        };


        private Builder() {
        }

        public Builder setEmptyView(@LayoutRes int emptyView,
                                    @Nullable CreateViewHolder creator) {
            this.emptyView = emptyView;
            return this;
        }

        public Builder setLoadingView(@LayoutRes int loadingView,
                                      @Nullable CreateViewHolder creator) {
            this.loadingView = loadingView;
            return this;
        }

        public Builder setFailureView(@LayoutRes int failureView,
                                      @Nullable CreateViewHolder creator) {
            this.failureView = failureView;
            return this;
        }

        public Builder setNextView(@LayoutRes int nextView,
                                   @Nullable CreateViewHolder creator) {
            this.nextView = nextView;
            return this;
        }

        public Builder setNextFailureView(@LayoutRes int nextFailureView,
                                          @Nullable CreateViewHolder creator) {
            this.nextFailureView = nextFailureView;
            return this;
        }

        public Builder setLoadListener(OnLoadListener loadListener) {
            this.loadListener = loadListener;
            return this;
        }

        public Builder addItemType(Class c, int layout, CreateViewHolder create) {
            typeArray.put(base, new MultiType(layout, create));
            typeMap.put(c.getName(), base);
            base++;
            return this;
        }

        public Builder setData(Object[] data) {
            this.data = data;
            return this;
        }

        public CommonRVAdapter build() {
            addStateType();
            return new CommonRVAdapter(this);
        }

        private void checkAndSetDefault() {
            if (emptyView <= 0) {
                emptyView = R.layout.item_empty;
            }
            if (loadingView <= 0) {
                loadingView = R.layout.item_loading;
            }
            if (failureView <= 0) {
                failureView = R.layout.item_loading_failure;
            }
            if (nextView <= 0) {
                nextView = R.layout.item_loading;
            }
            if (nextFailureView <= 0) {
                nextFailureView = R.layout.item_loading_failure;
            }
            if (emptyCreator == null) {
                emptyCreator = defaultCreator;
            }
            if (loadingCreator == null) {
                loadingCreator = defaultCreator;
            }
            if (failureCreator == null) {
                failureCreator = defaultCreator;
            }
            if (nextCreator == null) {
                nextCreator = defaultCreator;
            }
            if (nextFailureCreator == null) {
                nextFailureCreator = defaultCreator;
            }
        }

        private void addStateType() {
            checkAndSetDefault();
            typeArray.put(EMPTY, new MultiType(emptyView, emptyCreator));
            typeArray.put(LOADING, new MultiType(loadingView, loadingCreator));
            typeArray.put(LOADING_FAILURE, new MultiType(failureView, failureCreator));
            typeArray.put(LOADING_NEXT, new MultiType(nextView, nextCreator));
            typeArray.put(LOADING_NEXT_FAILURE, new MultiType(nextFailureView, nextFailureCreator));
        }
    }

}

File: app/src/main/java/com/nutrition/express/common/CommonRVAdapter.kt
package com.nutrition.express.common

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.nutrition.express.R
import java.util.*
import kotlin.collections.ArrayList
import kotlin.reflect.KClass


private const val TYPE_UNKNOWN = 0 //未知类型

private const val EMPTY = 10 /*显示EMPTY VIEW*/
private const val LOADING = 11 /*显示LOADING VIEW*/
private const val LOADING_FAILURE = 12 /*显示LOADING FAILURE VIEW*/
private const val LOADING_NEXT = 13 /*显示LOADING NEXT VIEW*/
private const val LOADING_NEXT_FAILURE = 14 /*显示LOADING NEXT FAILURE VIEW*/
private const val LOADING_FINISH = 15 //显示LOADING FINISH VIEW

class CommonRVAdapter private constructor(builder: Builder) :
    RecyclerView.Adapter<CommonViewHolder<Any>>() {

    /* 状态 */

    private var state = LOADING_FINISH

    private var isFinishViewEnabled = false
    private var isEmptyViewEnabled = false
    private var isLoadingViewEnabled = false
    private val loadListener: OnLoadListener?
    private val data: MutableList<Any>
    private val itemFactory: ItemViewHolderFactory
    private val stateFactory: StateViewHolderFactory

    private var extra: Any? = null

    private var onRetryListener: View.OnClickListener

    init {
        isFinishViewEnabled = builder.isFinishViewEnabled
        isEmptyViewEnabled = builder.isEmptyViewEnabled
        isLoadingViewEnabled = builder.isLoadingViewEnabled
        loadListener = builder.loadListener
        data = builder.data ?: ArrayList()
        itemFactory = builder.itemFactory
        stateFactory = builder.stateFactory

        onRetryListener = View.OnClickListener {
            loadListener?.let {
                it.retry()
                state = if (data.isNotEmpty()) LOADING_NEXT else LOADING
                notifyItemChanged(data.size)
            }
        }
        if (builder.data == null) {
            state = LOADING
        } else if (data.isEmpty()) {
            state = EMPTY
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CommonViewHolder<Any> {
        val vh = if (viewType in 10..15)
            stateFactory.createViewHolder(parent, viewType)
        else
            itemFactory.createViewHolder(parent, viewType)
        return if (vh == null) {     //check if unknown type
            val inflater = LayoutInflater.from(parent.context)
            val textView = inflater
                .inflate(android.R.layout.simple_list_item_1, parent, false) as TextView
            textView.text = "Error!!! Unknown type!!!"
            textView.setTextColor(Color.parseColor("#ff0000"))
            CommonViewHolder(textView)
        } else {
            if (viewType == LOADING_FAILURE) {
                vh.itemView.setOnClickListener(onRetryListener)
            } else if (viewType == LOADING_NEXT_FAILURE) {
                vh.itemView.setOnClickListener(onRetryListener)
            }
            vh as CommonViewHolder<Any>
        }
    }

    override fun onViewAttachedToWindow(holder: CommonViewHolder<Any>) {
        holder.onViewAttachedToWindow()
    }

    override fun onViewDetachedFromWindow(holder: CommonViewHolder<Any>) {
        holder.onViewDetachedFromWindow()
    }

    override fun onBindViewHolder(holder: CommonViewHolder<Any>, position: Int) {
        bindViewHolder(holder, position, null)
    }

    override fun onBindViewHolder(
        holder: CommonViewHolder<Any>,
        position: Int,
        payloads: List<Any>
    ) {
        bindViewHolder(holder, position, payloads)
    }

    private fun bindViewHolder(holder: CommonViewHolder<Any>, position: Int, payloads: List<Any>?) {
        if (position < data.size) {
            if (payloads.isNullOrEmpty()) {
                holder.bindView(data[position])
            } else {
                holder.bindView(data[position], payloads)
            }
        } else {
            holder.bindView(extra ?: "")
            if (state == LOADING_NEXT) {
                loadListener?.loadNextPage()
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == data.size) {
            state
        } else {
            itemFactory.getItemViewType(data[position]) ?: TYPE_UNKNOWN
        }
    }

    override fun getItemCount(): Int {
        if (data.size > 0) {
            if (state == EMPTY) {
                state = LOADING_FINISH
            }
        } else if (state == LOADING_FINISH) {
            state = EMPTY
        }
        return if (state == EMPTY && !isEmptyViewEnabled || state == LOADING_FINISH && !isFinishViewEnabled) {
            data.size
        } else {
            data.size + 1
        }
    }

    fun size(): Int {
        return data.size
    }

    fun getData(): List<Any> {
        return data
    }

    fun append(array: Array<Any>?, autoLoadingNext: Boolean) {
        if (array.isNullOrEmpty()) {
            if (data.size > 0) {
                //all data has been loaded
                state = LOADING_FINISH
                if (isFinishViewEnabled) {
                    notifyItemChanged(this.data.size)
                } else {
                    notifyItemRemoved(this.data.size)
                }
            } else {
                //no data, show empty view
                state = EMPTY
                notifyDataSetChanged()
            }
        } else {
            val lastDataIndex = data.size
            var size = array.size
            data.addAll(array)
            if (autoLoadingNext) {
                state = LOADING_NEXT
                size++
            } else {
                state = LOADING_FINISH
            }
            if (lastDataIndex == 0) {
                notifyDataSetChanged()
            } else {
                notifyItemRangeChanged(lastDataIndex, size)
            }
        }
    }

    fun remove(position: Int) {
        data.remove(position)
        notifyItemRemoved(position)
    }

    fun remove(start: Int, count: Int) {
        data.subList(start, start + count).clear()
        notifyItemRangeRemoved(start, count)
    }

    fun insert(start: Int, array: Array<Any>) {
        data.addAll(start, array.toMutableList())
        notifyItemRangeInserted(start, array.size)
    }

    fun move(fromPosition: Int, toPosition: Int) {
        Collections.swap(data, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    fun replace(index: Int, any: Any) {
        if (index >= 0 && index < data.size) {
            data[index] = any
            notifyItemChanged(index)
        }
    }

    fun clear() {
        data.clear()
        state = if (isLoadingViewEnabled) LOADING else EMPTY
        notifyDataSetChanged()
    }

    fun resetData(data: Array<Any>?, autoLoadingNext: Boolean) {
        this.data.clear()
        append(data, autoLoadingNext)
    }

    fun showLoadingFinish() {
        state = LOADING_FINISH
        notifyItemChanged(data.size)
    }

    fun showLoadingFailure(error: Any?) {
        extra = error
        state = if (data.isEmpty()) LOADING_FAILURE else LOADING_NEXT_FAILURE
        notifyItemChanged(data.size)
    }

    fun showEmptyView() {
        isEmptyViewEnabled = true
        state = EMPTY
        data.clear()
        notifyDataSetChanged()
    }

    fun showLoadingView() {
        state = LOADING
        notifyDataSetChanged()
    }

    fun showReloading() {
        state = LOADING
        data.clear()
        notifyDataSetChanged()
    }

    interface OnLoadListener {
        fun retry()
        fun loadNextPage()
    }

    interface ItemViewHolderFactory {
        fun getItemViewType(data: Any): Int? //the return viewType should not less than 1000
        fun createViewHolder(parent: ViewGroup, viewType: Int): CommonViewHolder<*>?
    }

    open class StateViewHolderFactory {
        fun createViewHolder(parent: ViewGroup, viewType: Int): CommonViewHolder<*>? {
            val inflater = LayoutInflater.from(parent.context)
            return when (viewType) {
                EMPTY ->
                    createEmptyView(inflater, parent)
                LOADING_FINISH ->
                    createFinishView(inflater, parent)
                LOADING, LOADING_NEXT ->
                    createLoadingView(inflater, parent)
                LOADING_FAILURE, LOADING_NEXT_FAILURE ->
                    createFailureView(inflater, parent)
                else -> null
            }
        }

        open fun createEmptyView(
            inflater: LayoutInflater,
            parent: ViewGroup
        ): CommonViewHolder<Any> {
            return CommonViewHolder(inflater.inflate(R.layout.item_empty, parent, false))
        }

        open fun createFinishView(
            inflater: LayoutInflater,
            parent: ViewGroup
        ): CommonViewHolder<Any> {
            return CommonViewHolder(inflater.inflate(R.layout.item_finish, parent, false))
        }

        open fun createLoadingView(
            inflater: LayoutInflater,
            parent: ViewGroup
        ): CommonViewHolder<Any> {
            return CommonViewHolder(inflater.inflate(R.layout.item_loading, parent, false))
        }

        open fun createFailureView(
            inflater: LayoutInflater,
            parent: ViewGroup
        ): CommonViewHolder<Any> {
            return ErrorViewHolder(inflater.inflate(R.layout.item_loading_failure, parent, false))
        }
    }

    companion object {
        private val defaultStateFactory = StateViewHolderFactory()
        private const val base = 1000

        fun adapter(init: Builder.() -> Unit): CommonRVAdapter {
            val builder = Builder()
            builder.init()
            return CommonRVAdapter(builder)
        }
    }

    class Builder {
        private val classList = ArrayList<KClass<*>>()
        private val layoutList = ArrayList<Int>()
        private val viewHolderList = ArrayList<(View) -> CommonViewHolder<*>>()

        var isFinishViewEnabled = false
        var isEmptyViewEnabled = false
        var isLoadingViewEnabled = false
        var loadListener: OnLoadListener? = null
        var data: MutableList<Any>? = null
        var stateFactory: StateViewHolderFactory = defaultStateFactory
        val itemFactory: ItemViewHolderFactory = object : ItemViewHolderFactory {
            override fun getItemViewType(data: Any): Int? {
                for (i in classList.indices) {
                    if (classList[i].isInstance(data)) {
                        return base + i
                    }
                }
                return null
            }

            override fun createViewHolder(parent: ViewGroup, viewType: Int): CommonViewHolder<*>? {
                val index = viewType - base
                if (index in 0 until layoutList.size) {
                    val view = LayoutInflater.from(parent.context)
                        .inflate(layoutList[index], parent, false)
                    return viewHolderList[index].invoke(view)
                }
                return null
            }
        }

        fun addViewType(clazz: KClass<*>, layout: Int, creator: (View) -> CommonViewHolder<*>) {
            classList.add(clazz)
            layoutList.add(layout)
            viewHolderList.add(creator)
        }

    }
}
